test/Modules/UngradedModules.jl

```diff
@@ -0,0 +1,675 @@
+		#######################################################################
+
+		# test, if H is surjective:
+		@test ImH == M
+
+#=
+		#3) H:N --> M neither injective nor surjective, also: tests 'restrict_domain()'
+		MM = Oscar.SubQuo(A1,B1)
+		M, iM, iSQ = Oscar.sum(MM, Oscar.SubQuo(A2,B1))
+		NN, p, i = Oscar.direct_product(MM,Oscar.SubQuo(A2,B2), task = :both)
+		i1,i2 = i[1],i[2]
+		p1,p2 = p[1],p[2]
+		nn = Oscar.ngens(NN)
+		N,iN = Oscar.sub(NN,[NN(Oscar.sparse_row(array_to_matrix([randpoly(R,0:15,2,1) for i=1:1,j=1:nn]))), NN(Oscar.sparse_row(array_to_matrix([randpoly(R,0:15,2,1) for i=1:1,j=1:nn]))), NN(Oscar.sparse_row(array_to_matrix([randpoly(R,0:15,2,1) for i=1:1,j=1:nn])))], :store)
+		
+		H = Oscar.restrict_domain(p1*iM,N)
+		@test Oscar.iswelldefined(H)
+
+		## testing the homomorphism theorem: #################################
+		KerH,iKerH = Oscar.kernel(H)
+		ImH,iImH = Oscar.image(H)
+
+		NmodKerH, pNmodKerH = Oscar.quo(N,KerH, :store)
+		Hbar = Oscar.SubQuoHom(NmodKerH,M,matrix(H)) # induced map N/KerH --> M
+		Hbar = Oscar.restrict_codomain(Hbar,ImH) # induced map N/KerH --> ImH
+
+		@test Oscar.iswelldefined(Hbar)
+		@test Oscar.isbijective(Hbar)
+
+		Hbar_inv = Oscar.inv(Hbar)
+
+		# test, if caching of inverse maps works:
+		@test Hbar_inv === Hbar.inverse_isomorphism
+		@test Hbar === Hbar_inv.inverse_isomorphism
+
+		# test if Hbar and Hbar_inv are inverse to each other:
+		@test all([Hbar_inv(Hbar(g))==g for g in Oscar.gens(NmodKerH)])
+		@test all([Hbar(Hbar_inv(g))==g for g in Oscar.gens(ImH)])
+		#######################################################################
+
+
+
+		#4) H: M --> N random map created via the hom() function
+		N = Oscar.SubQuo(A1,B1)
+		M = Oscar.SubQuo(A2,B2)
+		HomNM = Oscar.hom(N,M)[1]
+		H = HomNM(Oscar.sparse_row(array_to_matrix([randpoly(R,0:15,2,1) for _=1:1,j=1:Oscar.ngens(HomNM)])))
+		H = Oscar.homomorphism(H)
+		@test Oscar.iswelldefined(H)
+
+		## testing the homomorphism theorem: #################################
+		KerH,iKerH = Oscar.kernel(H)
+		ImH,iImH = Oscar.image(H)
+
+		NmodKerH, pNmodKerH = Oscar.quo(N,KerH, :store)
+		Hbar = Oscar.SubQuoHom(NmodKerH,M,matrix(H)) # induced map N/KerH --> M
+		#Hbar = AbstractAlgebra.Generic.ModuleHomomorphism(NmodKerH,M,matrix(H)) # induced map N/KerH --> M
+		Hbar = Oscar.restrict_codomain(Hbar,ImH) # induced map N/KerH --> ImH
+
+		@test Oscar.iswelldefined(Hbar)
+		@test Oscar.isbijective(Hbar)
+
+		Hbar_inv = Oscar.inv(Hbar)
+
+		# test, if caching of inverse maps works:
+		@test Hbar_inv === Hbar.inverse_isomorphism
+		@test Hbar === Hbar_inv.inverse_isomorphism
+
+		# test if Hbar and Hbar_inv are inverse to each other:
+		@test all([Hbar_inv(Hbar(g))==g for g in Oscar.gens(NmodKerH)])
+		@test all([Hbar(Hbar_inv(g))==g for g in Oscar.gens(ImH)])=#
+		#######################################################################
+	end
+	print("\r                                        ")
+	print("\r")
+end
```

Comment:
```suggestion
end

```
Please always end your files with a trailing newline.